src/index.ts

```diff
@@ -683,42 +684,30 @@ class Logging {
   getSinks(options?: GetSinksRequest): Promise<GetSinksResponse>;
   getSinks(callback: GetSinksCallback): void;
   getSinks(options: GetSinksRequest, callback: GetSinksCallback): void;
-  getSinks(
-      options?: GetSinksRequest|GetSinksCallback,
-      callback?: GetSinksCallback): void|Promise<GetSinksResponse> {
-    const self = this;
-    if (typeof options === 'function') {
-      callback = options as GetSinksCallback;
-      options = {};
-    }
+  async getSinks(opts?: GetSinksRequest|
+                 GetSinksCallback): Promise<GetSinksResponse> {
+    const options = opts ? opts as GetSinksRequest : {};
+    this.projectId = await this.auth.getProjectId();
```

Comment:
In an `async` method (like `getSinks`) I could just add a check 
```ts
if(!this.projectId) {
  this.projectId = awair this.auth.getProjectId();
}
```
In `getSinksStream` method however, that would be
```ts
userStream.once('reading', () => {
  if(!this.projectId) {
    this.auth.getProjectId().then(projectId => {
      this.projectId = projectId;
      const reqOpts = //... rest of the code
  })} else {
        const reqOpts = //... rest of the code
      }
});
```

I was a bit hesitant to duplicate code just like that.

Reply:
We also need to check that it's not the token value, so it gets even worse. But! I *think* that the auth client does that. Because, if it's working right, we give the auth client the `projectId` the user gave us. So, the client already knows not to try to detect one, it just returns what the user gave us.

Reply:
wouldn't bet about the auth client, looks like it is just checking for the token

https://github.com/googleapis/nodejs-projectify/blob/0e918a0b3a1f0fc9f99c9d986bb7cf33b3b84c33/src/index.ts#L44-L45

If we add a check for `'undefined'`, then we can make `projectId` in constructor optional and `replaceProjectIdToken` will still pick both (`'{{projectId}}'` or `'undefined'`) up.

Reply:
This is the part I was thinking about, that already knows what the provided project ID was: https://github.com/googleapis/google-auth-library-nodejs/blob/8fb65eb913cffa84e0557992f18f75c2a3d2d893/src/auth/googleauth.ts#L161

It will default to user-provided, then start rolling through the different best-guesses. That should mean we don't need to do any prior work, we can just call `this.auth.getProjectId()` and it will return what it should.